mithril-common/src/crypto_helper/merkle_map.rs

```diff
@@ -0,0 +1,617 @@
+            MKMapNode::Tree(merkle_tree) => merkle_tree.contains(&leaf),
+            MKMapNode::TreeNode(merkle_tree_node) => *merkle_tree_node == leaf,
+        }
+    }
+
+    fn compute_proof<T: Into<MKTreeNode> + Clone>(
+        &self,
+        leaves: &[T],
+    ) -> StdResult<Option<MKMapProof<K>>> {
+        match self {
+            MKMapNode::Tree(ref value) => {
+                let proof = value
+                    .compute_proof(
+                        &leaves
+                            .iter()
+                            .map(|leaf| leaf.to_owned().into())
+                            .collect::<Vec<_>>(),
+                    )
+                    .with_context(|| "MKMapValue could not compute sub proof for MKTree")?;
+                Ok(Some(proof.into()))
+            }
+            MKMapNode::Map(ref value) => {
+                let proof = value
+                    .compute_proof(
+                        &leaves
+                            .iter()
+                            .map(|leaf| leaf.to_owned().into())
+                            .collect::<Vec<_>>(),
+                    )
+                    .with_context(|| "MKMapValue could not compute sub proof for MKMap")?;
+                Ok(Some(proof))
+            }
+            _ => Ok(None),
+        }
+    }
+}
+
+impl<K: MKMapKey> From<MKMap<K, MKMapNode<K>>> for MKMapNode<K> {
+    fn from(other: MKMap<K, MKMapNode<K>>) -> Self {
+        MKMapNode::Map(Rc::new(other))
+    }
+}
+
+impl<K: MKMapKey> From<MKTree> for MKMapNode<K> {
+    fn from(other: MKTree) -> Self {
+        MKMapNode::Tree(Rc::new(other))
+    }
+}
+
+impl<K: MKMapKey> From<MKTreeNode> for MKMapNode<K> {
+    fn from(other: MKTreeNode) -> Self {
+        MKMapNode::TreeNode(other)
+    }
+}
+
+impl<K: MKMapKey> TryFrom<MKMapNode<K>> for MKTreeNode {
+    type Error = StdError;
+    fn try_from(other: MKMapNode<K>) -> Result<Self, Self::Error> {
+        other.compute_root()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::collections::BTreeSet;
+    use std::ops::Range;
+
+    use crate::entities::BlockRange;
+
+    use super::*;
+
+    fn generate_merkle_trees(
+        total_leaves: u64,
+        block_range_length: u64,
+    ) -> Vec<(BlockRange, MKTree)> {
+        (0..total_leaves / block_range_length)
+            .map(|block_range_index| {
+                let block_range = BlockRange::new(
+                    block_range_index * block_range_length,
+                    (block_range_index + 1) * block_range_length,
+                );
+                let leaves = <Range<u64> as Clone>::clone(&block_range)
+                    .map(|leaf_index| leaf_index.to_string())
+                    .collect::<Vec<_>>();
+                let merkle_tree_block_range = MKTree::new(&leaves).unwrap();
+
+                (block_range, merkle_tree_block_range)
+            })
+            .collect::<Vec<_>>()
+    }
+
+    #[test]
+    fn test_mk_map_should_compute_consistent_root() {
```

Comment:
I have trouble to understand what is tested here.
If it's only the equivalence between `MKMapNode::TreeNode` and a `MKTree` then the test could be simplified to directly test the the `MKMapNode::compute_root` result is equal to the `MKTree::compute_root` it derives from (you don't need to generate MKMap to test that).

Reply:
It's testing the fact that if when you replace an entry with another entry that reduces to the same root, the computation of the map root is the same.

Reply:
So the test should be named approximately like this: `when_replacing_a_node_by_equivalent_tree_compute_the_same_root` ?